drivers/feetech/feetech.c

```diff
@@ -0,0 +1,129 @@
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     drivers_feetech
+ * @{
+ *
+ * @file
+ * @brief       Driver implementation for Feetech devices
+ *
+ * @author      Loïc Dauphin <loic.dauphin@inria.fr>
+ *
+ * @}
+ */
+
+#include "feetech.h"
+
+#include "feetech_protocol.h"
+#include "feetech_reader.h"
+#include "feetech_writer.h"
+
+#include "periph/uart.h"
+#include "xtimer.h"
+
+#include <string.h>
+
+static inline void _n2h(uint16_t* value)
+{
+    *value = (((*value) >> 8) & 0xFF) | (((*value) & 0xFF) << 8);
+}
+
+bool feetech_ping(uart_half_duplex_t *stream, feetech_id_t id)
+{
+    feetech_writer_t pw;
+
+    uart_half_duplex_set_tx(stream);
+    feetech_writer_init(&pw, stream->buffer, stream->size);
+    feetech_writer_ping_make(&pw, id);
+    uart_half_duplex_send(stream, pw.size);
+
+    uart_half_duplex_set_rx(stream);
+    return uart_half_duplex_recv(stream, FEETECH_ACK_SIZE) == FEETECH_ACK_SIZE;
+}
+
+void feetech_init(feetech_t *device, uart_half_duplex_t *stream, feetech_id_t id)
+{
+    device->stream = stream;
+    device->id = id;
+}
+
+bool feetech_write(feetech_t *device, feetech_addr_t reg, const uint8_t *data, size_t length)
+{
+    if (device->stream->size < length) {
+        return false;
+    }
+
+    feetech_writer_t pw;
+
+    uart_half_duplex_set_tx(device->stream);
+    feetech_writer_init(&pw, device->stream->buffer, device->stream->size);
+    feetech_writer_write_make(&pw, device->id, reg, data, length);
+    uart_half_duplex_send(device->stream, pw.size);
+
+    uart_half_duplex_set_rx(device->stream);
+    return uart_half_duplex_recv(device->stream, FEETECH_ACK_SIZE) == FEETECH_ACK_SIZE;
+}
+
+bool feetech_write8(feetech_t *device, feetech_addr_t reg, uint8_t value)
+{
+    return feetech_write(device, reg, &value, 1);
+}
+
+bool feetech_write16(feetech_t *device, feetech_addr_t reg, uint16_t value)
+{
+    _n2h(&value);
+    return feetech_write(device, reg, (uint8_t*)&value, 2);
+}
+
+bool feetech_read(feetech_t *device, feetech_addr_t reg, uint8_t *data, size_t length)
+{
+    if (device->stream->size < length) {
+        return false;
+    }
+
+    feetech_writer_t pw;
+
+    uart_half_duplex_set_tx(device->stream);
+    feetech_writer_init(&pw, device->stream->buffer, device->stream->size);
+    feetech_writer_read_make(&pw, device->id, reg, length);
+    uart_half_duplex_send(device->stream, pw.size);
+
+    uart_half_duplex_set_rx(device->stream);
+    const size_t esize = FEETECH_RESPONSE_SIZE(length);
+    if (uart_half_duplex_recv(device->stream, esize) != esize) {
+        return false;
+    }
+
+    feetech_reader_t pr;
+    feetech_reader_init(&pr, device->stream->buffer, esize);
+    if (!feetech_reader_is_valid(&pr)) {
+        return false;
+    }
+
+    if (feetech_reader_response_get_payload_size(&pr) != length) {
```

Comment:
you are calling the same function twice. Would it make sense to use something like: 
```c
    }
+
+    if (feetech_reader_response_get_payload_size(&pr, data) != length) {
    return false;
}

return true;
```